src/common/str_util.h

```diff
@@ -0,0 +1,289 @@
+    return true;
+  } else {
+    return false;
+  }
+}
+
+
+// Same thing with a C-style array
+template<std::size_t N>
+bool nul_terminated_copy(std::string_view source,
+			 char (&dest)[N])
+{
+  if constexpr (N == 0) {
+    return true;
+  }
+  return nul_terminated_copy<N>(source, &dest[0]);
+}
+
+// Same thing with a standard array
+template<std::size_t N>
+bool nul_terminated_copy(std::string_view source,
+			 std::array<char, N>& dest)
+{
+  if constexpr (N == 0) {
+    return true;
+  }
+  return nul_terminated_copy<N>(source, dest.data());
+}
+
+// Same as above, except with RVO instead of caller-provided-storage.
+//
+// If the total length of source is less than N, returns an optional
+// array containing a NUL-terminated copy.
+//
+// Otherwise, returns an empty optional.
+template<std::size_t N>
+std::optional<std::array<char, N>>
+nul_terminated_copy(std::string_view source)
+{
+  std::optional<std::array<char, N>> dest(std::in_place);
+  auto b = nul_terminated_copy<N>(source, dest->data());
+  if (!b)
+    dest.reset();
+  return dest;
+}
+
+// Control Flow results, to control breaking substr_do (and possibly
+// other things down the road.)
+enum class cf {
+  go, // Continue execution
+  stop // Stop here
+};
+
+// Default delimiters for substr_do and associates.
+inline constexpr std::string_view default_delims = ";,= \t";
+inline constexpr std::string_view space = " \f\n\r\t\v";
+
+/// Split a string using the given delimiters, passing each piece as a
+/// (non-null-terminated) std::string_view to the callback.
+template<typename F>
+void substr_do(
+  std::string_view s, F&& f,
+  std::enable_if_t<!std::is_same_v<std::invoke_result_t<F, std::string_view>,
+                                   cf>, std::string_view> d = default_delims)
+{
+  auto pos = s.find_first_not_of(d);
+  while (pos != s.npos) {
+    s.remove_prefix(pos);
+    auto end = s.find_first_of(d);
+    f(s.substr(0, end));
+    pos = s.find_first_not_of(d, end);
+  }
+}
+
+// As above, but let the function specify whether to continue after
+// every call. f, must return a member of cf at every
+// execution. cf::stop will, unsurprisingly, stop.
+template<typename F>
+void substr_do(
+  std::string_view s, F&& f,
+  std::enable_if_t<std::is_same_v<std::invoke_result_t<F, std::string_view>,
+                                  cf>, std::string_view> d = default_delims)
+{
+  auto pos = s.find_first_not_of(d);
+  while (pos != s.npos) {
+    s.remove_prefix(pos);
+    auto end = s.find_first_of(d);
+    cf b = f(s.substr(0, end));
+    if (b == cf::stop)
+      break;
+    pos = s.find_first_not_of(d, end);
+  }
+}
+
+// Insert each delimited substring into a container, specified by the
+// output iterator. Will insert a std::string_view if possible, and a
+// std::string otherwise.
+template<typename Iterator>
+Iterator substr_insert(std::string_view s,
+		       Iterator i,
+		       std::string_view d = default_delims)
+{
+  substr_do(s,
+	    [&](auto&& s) {
```

Comment:
is there a reason for callers to prefer `auto&&` here instead of `std::string_view`?

Reply:
It's shorter. No other reason. Unreferenced auto would probably work just as well since string_view is only two words.